### PR TITLE
[dy] Escape password in db_connection_url

### PR DESCRIPTION
**Pull request by dy46**
# Summary

We currently don't escape characters in the db_connection_url, but the sqlalchemy and alembic docs do suggest doing so. I also ran into an error when I had a special character in the password which caused the alembic migrations to fail.

Docs:
https://docs.sqlalchemy.org/en/14/core/engines.html#database-urls
https://alembic.sqlalchemy.org/en/latest/api/config.html#alembic.config.Config.set_main_option

# Tests

tested locally with postgres and sqlite

cc: @wangxiaoyou1993

### CODE REVIEW

**Comment by wangxiaoyou1993** [comment]
Could you give an example what db_connection_url has the issue?

**Comment by dy46** [comment]
soemthing like this
postgresql+psycopg2://postgres:abc%3def$123@xxx.us-west-2.rds.amazonaws.com:5432/postgres

**Comment by wangxiaoyou1993** [comment]
> soemthing like this postgresql+psycopg2://postgres:abc%3def$[123@xxx.us-west-2.rds.amazonaws.com](mailto:123@xxx.us-west-2.rds.amazonaws.com):5432/postgres

Thanks. Just wondering, could it break things if user already provides an escaped url in the db_connection_url?

**Comment by dy46** [comment]
that is a good point

**Comment by dy46** [comment]
@wangxiaoyou1993 updated the PR to escape the strings only when certain exceptions are raised

File: mage_ai/orchestration/db/__init__.py [hunk 1]
```diff
@@ -51,10 +52,25 @@
...
+    engine.connect()
+except SQLAlchemyError:
+    engine.dispose()
+    url_parsed = urlparse(db_connection_url)
+    if url_parsed.password:
+        db_connection_url = db_connection_url.replace(
+            url_parsed.password,
+            quote_plus(url_parsed.password),
+        )
+    engine = create_engine(
```
Comment (wangxiaoyou1993) [inline mage_ai/orchestration/db/__init__.py:69]: If there's no url_parsed.password, we can directly throw exception, right?

**Review by wangxiaoyou1993 — APPROVED** [review]

File: mage_ai/orchestration/db/__init__.py (hunk 1 above)
Comment (dy46) [reply mage_ai/orchestration/db/__init__.py:69]: oops i didn't see this, i'll add it as a follow up

**Pull request merged**